File: 198. House Robber.java
import java.util.*;
class Solution {
    
    // hash table to store refs
    // key: length of sub array, val: maximum amount of money to rob
    Hashtable<Integer, Integer> t = new Hashtable<>();
    
    public int rob(int[] nums) {
        
        int n = nums.length;
        
        // base cases
        if (n == 0) {
            return 0;
        }
        if (n == 1) {
            return nums[0];
        }
        
        // check hash table for cache
        if (t.containsKey(n)) {
            return t.get(n);
        }
        
        // choose to rob the first house
        int r1 = nums[0] + rob(Arrays.copyOfRange(nums, 2, n));

        // choose not to rob
        int r2 = rob(Arrays.copyOfRange(nums, 1, n));
        
        // save res in hash table
        t.put(n, Math.max(r1, r2));
        
        // return 
        return t.get(n);
        
        
    }
}

File: 198. House Robber.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        
        # key: house idx in nums
        # val: max amount to rob
        t = {}
        
        n = len(nums)
        
        
        def helper(idx):
            
            # base case: idx out of range
            if idx >= n:
                return 0
            
            # ref the hash table
            if idx in t:
                return t[idx]
            
            # for each house there are 2 options

            # rob (and therefore can't rob the next one)
            m1 = nums[idx] + helper(idx + 2)

            # not rob
            m2 = helper(idx + 1)
            
            # record in hash table
            t[idx] = max(m1, m2)
            
            return t[idx]
        
        # call helper to calculate
        return helper(0)